Mathlib/CategoryTheory/Limits/Shapes/Terminal.lean

```diff
@@ -702,50 +712,64 @@ def colimitOfDiagramInitial {X : J} (hX : IsInitial X) (F : J ⥤ C)
   desc S := S.ι.app _
 #align category_theory.limits.colimit_of_diagram_initial CategoryTheory.Limits.colimitOfDiagramInitial
 
+instance hasColimit_of_domain_hasInitial [HasInitial J] {F : J ⥤ C}
+    [∀ (i j : J) (f : i ⟶ j), IsIso (F.map f)] : HasColimit F :=
+  HasColimit.mk { cocone := _, isColimit := colimitOfDiagramInitial (initialIsInitial) F }
+
 -- This is reducible to allow usage of lemmas about `cocone_point_unique_up_to_iso`.
 /-- For a functor `F : J ⥤ C`, if `J` has an initial object and all the morphisms in the diagram
 are isomorphisms, then the image of the initial object is isomorphic to the colimit of `F`. -/
 @[reducible]
-def colimitOfInitial (F : J ⥤ C) [HasInitial J] [HasColimit F]
-    [∀ (i j : J) (f : i ⟶ j), IsIso (F.map f)] : colimit F ≅ F.obj (⊥_ J) :=
+def colimitOfInitial (F : J ⥤ C) [HasInitial J] [∀ (i j : J) (f : i ⟶ j), IsIso (F.map f)] :
+    colimit F ≅ F.obj (⊥_ J) :=
   IsColimit.coconePointUniqueUpToIso (colimit.isColimit _)
     (colimitOfDiagramInitial initialIsInitial _)
 #align category_theory.limits.colimit_of_initial CategoryTheory.Limits.colimitOfInitial
 
 /-- If `j` is initial in the index category, then the map `limit.π F j` is an isomorphism.
 -/
-theorem isIso_π_of_isInitial {j : J} (I : IsInitial j) (F : J ⥤ C) [HasLimit F] :
+theorem isIso_π_of_isInitial {j : J} (I : IsInitial j) (F : J ⥤ C) :
+    haveI := IsInitial.hasInitial I
     IsIso (limit.π F j) :=
+  haveI := IsInitial.hasInitial I
   ⟨⟨limit.lift _ (coneOfDiagramInitial I F), ⟨by ext; simp, by simp⟩⟩⟩
 #align category_theory.limits.is_iso_π_of_is_initial CategoryTheory.Limits.isIso_π_of_isInitial
 
-instance isIso_π_initial [HasInitial J] (F : J ⥤ C) [HasLimit F] : IsIso (limit.π F (⊥_ J)) :=
+instance isIso_π_initial [HasInitial J] (F : J ⥤ C) : IsIso (limit.π F (⊥_ J)) :=
   isIso_π_of_isInitial initialIsInitial F
 #align category_theory.limits.is_iso_π_initial CategoryTheory.Limits.isIso_π_initial
 
-theorem isIso_π_of_isTerminal {j : J} (I : IsTerminal j) (F : J ⥤ C) [HasLimit F]
-    [∀ (i j : J) (f : i ⟶ j), IsIso (F.map f)] : IsIso (limit.π F j) :=
+theorem isIso_π_of_isTerminal {j : J} (I : IsTerminal j) (F : J ⥤ C)
+    [∀ (i j : J) (f : i ⟶ j), IsIso (F.map f)] :
+    haveI := IsTerminal.hasTerminal I
+    IsIso (limit.π F j) :=
+  haveI := IsTerminal.hasTerminal I
```

Comment:
For this statement (and the dual version below), I would keep the `[HasLimit F]` assumption.

Reply:
Any particular reason why? Should I then also revert the versions in lines 652 and 723 with similar statements,  and remove the instances lines 715 and 644?

Reply:
The only thing that I feel *could* be considered questionable in this PR is the `haveI` in the *statement* and the dual version. If we can refrain to do so, it is probably better: in my code, I would do this only if this is the only solution. Here, I would think a term like `limit.π F j` would appear in any definition or proof only if the instance `[HasLimit F]` can be found, so that it should not be harmful to assume `[HasLimit F]` here. Other maintainers/reviewers may have slightly different opinions though.

Reply:
Then, there are still `haveI` in two statements (lines 731 and 755). Should I also remove them? While I agree that in most situations the `HasLimit F` will be found somewhere anyway, I felt it was nice that providing an initial object proves in itself that `F` has a limit.

Reply:
Ah, yes, I would do the same for these two other statements.